### DISCUSSION TITLE
Middlewares and livewire actions

### DISCUSSION THREAD

**Discussion by pktharindu** [livewire/livewire#3067] [7. Bugs (archived)]
**Question:** How to add middleware to a Livewire action?

For example, I have the following in my application:

web route:
```php
Route::livewire('email/verify', 'auth.verify')
        ->layout('layouts.auth')
        ->middleware(['auth', 'throttle:6,1'])
        ->name('verification.notice');
```

livewire component:
```php
<?php

namespace App\Http\Livewire\Auth;

use Livewire\Component;
use Illuminate\Support\Facades\Auth;

class Verify extends Component
{
    public function resend()
    {
        if (Auth::user()->hasVerifiedEmail()) {
            redirect(route('home'));
        }

        Auth::user()->sendEmailVerificationNotification();

        $this->emit('resent');

        session()->flash('resent');
    }

    public function render()
    {
        return view('livewire.auth.verify');
    }
}
```

Adding the `throttle` middleware to the route like this also throttles the view itself. Obviously I only want to throttle the `resend` action. How can I achieve this?

**Comment by booni3** [comment]
Take the throttle middleware off the livewire route and dispatch the email via a job with the throttle on that.

**Comment by pktharindu** [comment]
Thanks @booni3. From what I can understand, it will throttle the job, but that won't return `429 Too Many Requests` error.
It will instead keep showing the flash message saying "verification link has been sent to your email" which isn't exactly true.

**Comment by booni3** [comment]
Getting response messages back from background jobs is a bit more complicated. You would need to push your flash message to the front end with websockets in this case.


Saying this now there is probably a simpler way with livewire.

I’ll have a think but someone else might have a suggestion.

**Comment by danharrin** [comment]
Hi, I'm gonna close this thread for now. Please feel free to use [the discussions section](https://github.com/livewire/livewire/discussions) or the [Discord server](https://discord.gg/H884eyt) to continue conversation about this issue. Thanks!

**Comment by mokhosh** [comment]
REOPEN

**Comment by mokhosh** [comment]
I think this is necessary to be able to use Livewire for things like Login, Code Verification, etc.

The ability to override middlewares per action.

**Comment by austenc** [comment]
Hey @mokhosh do you need middlewares _per action_, or would a _per component_ basis work well enough? 

The recent changes to middleware to make them more persistent should handle most use cases. If that's the case, can we close out this issue? 

See here: https://github.com/livewire/livewire/releases/tag/v2.3.9

Thank you!

**Comment by mokhosh** [comment]
@austenc I think I was looking for per action at the time.
I think @danharrin has created a package for per method throttling though.

Feel free to close, we can reopen or create a new issue if something comes up.

**Comment by danharrin** [comment]
@austenc @mokhosh Yeah, I built [this package for rate limiting actions](https://github.com/danharrin/livewire-rate-limiting). Would love it to be a core feature though!

**Comment by mokhosh** [comment]
I think other middlewares can be useful too. Like making empty strings null in a certain action, or applying a certain authorisation middleware, etc.

**Comment by pktharindu** [comment]
@danharrin's package solves the exact problem I had at the time. Thank you for this. 👌🏻💯

That being said, it would still be useful to have the flexibility to apply middleware per action.
